### PR TITLE
Merge with upstream changes

### PR DESCRIPTION
**Pull request by Thomas-Neill**
This should fix the segfault bugs we've been having.

We should merge https://github.com/HalloAppInc/gpb/pull/5 first.

ejabberd works when we use the merged versions of this and gpb

### CODE REVIEW

**Comment by vipin211** [comment]
Can you please rebase changes made specifically by you into the last commit and describe your changes in the commit description?

**Comment by Thomas-Neill** [comment]
> Can you please rebase changes made specifically by you into the last commit and describe your changes in the commit description?

Will do, didn't think about how this would be hard to review.

**Comment by vipin211** [comment]
> > Can you please rebase changes made specifically by you into the last commit and describe your changes in the commit description?
> 
> Will do, didn't think about how this would be hard to review.

Not hard to review, but for record keeping.

**Comment by Thomas-Neill** [comment]
@vipin211 i've squashed changes on both repos now

**Comment by vipin211** [comment]
will ping you tomorrow to review in person. sorry for the delay.

File: .github/workflows/ci.yml
Comment (kennedyjosh) [inline .github/workflows/ci.yml]: remove this

File: .github/workflows/ci.yml
Comment (kennedyjosh) [inline .github/workflows/ci.yml]: only need 23 and 24

File: .github/workflows/ci.yml
Comment (kennedyjosh) [inline .github/workflows/ci.yml]: we don't need to test this on mac

File: rebar.config
Comment (kennedyjosh) [inline rebar.config:13]: looks like this is here twice

**Review by kennedyjosh — APPROVED** [review]
Same as gpb pr, we should watch this closely on s-test for at least a day

File: rebar.config
Comment (Thomas-Neill) [reply rebar.config:13]: this was removed in the commit after i merged

File: .gitignore
Comment (loukiano) [inline .gitignore]: I'm not sure we should ignore `rebar.lock` given that it overrides `rebar.config` unless specifically upgraded, I could see this causing some unintentional problems

**Comment by balusu-murali** [comment]
> Same as gpb pr, we should watch this closely on s-test for at least a day

these kind of changes are hard to judge even in a single day.

File: c_src/enif_protobuf.h
Comment (loukiano) [inline c_src/enif_protobuf.h:44]: I kind of like these debuggers printing to stderr given that it's more reliably flushed. Personal preference but I would prefer keeping these as `fprintf(stderr...`

File: c_src/enif_protobuf.h
Comment (loukiano) [inline c_src/enif_protobuf.h:59]: Error/Exception functions should definitely be printing to stderr imo

File: c_src/enif_protobuf.h
Comment (loukiano) [inline c_src/enif_protobuf.h:65]: super small nit but do we really need `\r` in these? I'm pretty sure `\n` does both newline and carriage return in C

File: .gitignore
Comment (kennedyjosh) [reply .gitignore]: rebar specifically recommends not to ignore the rebar.lock file, good catch

File: c_src/enif_protobuf.h
Comment (loukiano) [reply c_src/enif_protobuf.h:65]: Thought more about this and I think it's better this way; forgot that escape sequences are interpreted by the output processor, not C.

File: c_src/ep_decoder.c
Comment (Thomas-Neill) [inline c_src/ep_decoder.c:914]: @vipin211 
investigated and for enums, `v_fields` is a copy of `fields` which is sorted by integer value assigned to the enum. `fields` is sorted arbitrarily by the IDs of the atoms used for the fields. as a result i think our behavior is correct here (default should be the min. enum value) and theirs isn't. i'll take a look at the spec and make a PR upstream.

**Review by vipin211 — APPROVED** [review]
I think we should try this code this week once we have various approvals.

File: c_src/ep_decoder.c
Comment (balusu-murali) [reply c_src/ep_decoder.c:914]: i made some fixes on our local copy to set the enum to the default field, to set bool default value as false etc.
https://github.com/HalloAppInc/enif_protobuf/commit/8848508735de20ea3287f436b5b70595ef307c85

why are we removing this code? dont think we should.
i am not sure about existing tests in this repo - but we should keep this code.

File: c_src/ep_encoder.c
Comment (balusu-murali) [inline c_src/ep_encoder.c:522]: why change this?

File: c_src/ep_encoder.c
Comment (balusu-murali) [inline c_src/ep_encoder.c:743]: did something change in these functions?

**Review by balusu-murali — CHANGES_REQUESTED** [review]

File: c_src/ep_decoder.c
Comment (balusu-murali) [inline c_src/ep_decoder.c:910]: i thought the previous code was also doing something similar.. what are you trying to do here?

File: c_src/ep_encoder.c
Comment (Thomas-Neill) [reply c_src/ep_encoder.c:522]: `enc->omit` was removed with the proto3 changes upstream; this is what they've changed this idiom to elsewhere in the file and this behavior seems correct to me.

File: c_src/ep_encoder.c
Comment (Thomas-Neill) [reply c_src/ep_encoder.c:743]: yes, i think they've changed the logic for when to not encode zero fields

File: c_src/ep_decoder.c
Comment (Thomas-Neill) [reply c_src/ep_decoder.c:910]: yes, this also should get the enum with a value of zero. the main difference is that this works if there's ever an enum assigned a negative value as you can't just grab the first one in that case.

File: c_src/ep_decoder.c
Comment (Thomas-Neill) [reply c_src/ep_decoder.c:914]: i think we should be going into the switch statement a few lines above to fill out default values now that that's been added; i forget what the proto compiler sets the occurrence options to by default, but if it's not already `repeated` or `defaulty` i can just remove the conditions for the above case in the if statement so that we get the same behavior.

File: c_src/ep_decoder.c
Comment (balusu-murali) [reply c_src/ep_decoder.c:910]: we should never compile if there was a negative value in the enum in the first place.
dont follow why we have to do a binary search here in the first place.

File: c_src/ep_encoder.c
Comment (balusu-murali) [reply c_src/ep_encoder.c:522]: did you get a chance to check if it works fine with the protobufs defined in our repo.
you can test some protobuf definitions in this repo standalone.. were you able to test our protobufs?

File: c_src/ep_decoder.c
Comment (Thomas-Neill) [reply c_src/ep_decoder.c:910]: it's technically legal but not recommended in the proto3 spec to have a negative value. i don't think we will ever be making a ton of enums with negative values, but when i made the pr upstream it's good to support that too.

File: c_src/ep_decoder.c
Comment (Thomas-Neill) [reply c_src/ep_decoder.c:914]: checked again and everything's marked as `optional` - i'll make the change to just fill everything with the default value

File: c_src/ep_encoder.c
Comment (Thomas-Neill) [reply c_src/ep_encoder.c:522]: ok looked into it and since `proto_v` is set to `2` by default and `enc->omit` was also only set if `proto_v` was `3`, i don't think this case wasn't going off before, so this shouldn't change anything. i'm still going to test this case to make sure it works the same as on master though.

Comment (Thomas-Neill) [reply c_src/ep_encoder.c:522]:
```
-record(m1, {a}).

encode_msg(Msg, Defs) ->
    encode_msg(Msg, Defs, [{string_as_list, true}, {with_utf8, true}]).

encode_msg(Msg, Defs, Opts) ->
    enif_protobuf:set_opts(Opts),
    enif_protobuf:load_cache(Defs),
    enif_protobuf:encode(Msg).


encodes_binary_correctly_test() ->
    <<8, 0>> = encode_msg( #m1{a = <<"0">>},
        [{{msg, m1}, [#?gpb_field{name = a, fnum = 1, rnum = #m1.a,
                type = int64, occurrence = optional,
                opts = [{ebin, true}]}]}]
        ).
```
This snippet passes on both `master` and this branch so I'm pretty confident this change isn't breaking anything.

File: c_src/ep_decoder.c
Comment (balusu-murali) [reply c_src/ep_decoder.c:910]: ah - you are right about negative values. let us please do this in a separate diff.
let us keep this diff to just pulling in latest changes from master.

File: c_src/ep_encoder.c
Comment (balusu-murali) [reply c_src/ep_encoder.c:522]: i would use our own protobuf definitions - are you sure the proto_v is being picked up properly from our protobuf files?
the default occurrence type is optional for proto3.

File: rebar.config
Comment (balusu-murali) [inline rebar.config:2]: dont think these are linked. get rid of gpb changes here - i would prefer we upgrade gpb completely separately.

File: c_src/ep_decoder.c
Comment (Thomas-Neill) [reply c_src/ep_decoder.c:910]: it has been merged on master upstream now

**Comment by Thomas-Neill** [comment]
closing - the other PR is cleaner

**Pull request closed without merging**
